File: android/src/main/kotlin/dev/note11/flutter_naver_map/flutter_naver_map/controller/overlay/handler/PathOverlayHandler.kt
package dev.note11.flutter_naver_map.flutter_naver_map.controller.overlay.handler

import com.naver.maps.map.overlay.Overlay
import com.naver.maps.map.overlay.PathOverlay
import dev.note11.flutter_naver_map.flutter_naver_map.controller.overlay.OverlayHandler
import dev.note11.flutter_naver_map.flutter_naver_map.controller.overlay.OverlayHandler.Companion.getterName
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.overlay.overlay.NPathOverlay
import io.flutter.plugin.common.MethodChannel

internal interface PathOverlayHandler : OverlayHandler {
    fun handlePathOverlay(
        pathOverlay: Overlay, method: String, arg: Any?, result: MethodChannel.Result,
    ) = (pathOverlay as PathOverlay).let { p ->
        when (method) {
            NPathOverlay.coordsName -> setCoords(p, arg!!)
            NPathOverlay.widthName -> setWidth(p, arg!!)
            NPathOverlay.colorName -> setColor(p, arg!!)
            NPathOverlay.outlineWidthName -> setOutlineWidth(p, arg!!)
            NPathOverlay.outlineColorName -> setOutlineColor(p, arg!!)
            NPathOverlay.passedColorName -> setPassedColor(p, arg!!)
            NPathOverlay.passedOutlineColorName -> setPassedOutlineColor(p, arg!!)
            NPathOverlay.progressName -> setProgress(p, arg!!)
            NPathOverlay.patternImageName -> setPatternImage(p, arg!!)
            NPathOverlay.patternIntervalName -> setPatternInterval(p, arg!!)
            NPathOverlay.isHideCollidedCaptionsName -> setIsHideCollidedCaptions(p, arg!!)
            NPathOverlay.isHideCollidedMarkersName -> setIsHideCollidedMarkers(p, arg!!)
            NPathOverlay.isHideCollidedSymbolsName -> setIsHideCollidedSymbols(p, arg!!)
            getterName(NPathOverlay.boundsName) -> getBounds(p, result::success)
            else -> result.notImplemented()
        }
    }

    fun setCoords(pathOverlay: PathOverlay, rawCoords: Any)

    fun setWidth(pathOverlay: PathOverlay, rawWidthDp: Any)

    fun setColor(pathOverlay: PathOverlay, rawColor: Any)

    fun setOutlineWidth(pathOverlay: PathOverlay, rawWidthDp: Any)

    fun setOutlineColor(pathOverlay: PathOverlay, rawColor: Any)

    fun setPassedColor(pathOverlay: PathOverlay, rawColor: Any)

    fun setPassedOutlineColor(pathOverlay: PathOverlay, rawColor: Any)

    fun setProgress(pathOverlay: PathOverlay, rawProgress: Any)

    fun setPatternImage(pathOverlay: PathOverlay, rawNOverlayImage: Any)

    fun setPatternInterval(pathOverlay: PathOverlay, rawInterval: Any)

    fun setIsHideCollidedCaptions(pathOverlay: PathOverlay, rawFlag: Any)

    fun setIsHideCollidedMarkers(pathOverlay: PathOverlay, rawFlag: Any)

    fun setIsHideCollidedSymbols(pathOverlay: PathOverlay, rawFlag: Any)

    fun getBounds(pathOverlay: PathOverlay, success: (bounds: Map<String, Any>) -> Unit)
}

File: ios/flutter_naver_map/Sources/flutter_naver_map/controller/overlay/handler/PathOverlayHandler.swift
import NMapsMap
import Flutter

internal protocol PathOverlayHandler: OverlayHandler {
    func setCoords(_ pathOverlay: NMFPath, rawCoords: Any)

    func setWidth(_ pathOverlay: NMFPath, rawWidthDp: Any)

    func setColor(_ pathOverlay: NMFPath, rawColor: Any)

    func setOutlineWidth(_ pathOverlay: NMFPath, rawWidthDp: Any)

    func setOutlineColor(_ pathOverlay: NMFPath, rawColor: Any)

    func setPassedColor(_ pathOverlay: NMFPath, rawColor: Any)

    func setPassedOutlineColor(_ pathOverlay: NMFPath, rawColor: Any)

    func setProgress(_ pathOverlay: NMFPath, rawProgress: Any)

    func setPatternImage(_ pathOverlay: NMFPath, rawNOverlayImage: Any)

    func setPatternInterval(_ pathOverlay: NMFPath, rawInterval: Any)

    func setIsHideCollidedCaptions(_ pathOverlay: NMFPath, rawFlag: Any)

    func setIsHideCollidedMarkers(_ pathOverlay: NMFPath, rawFlag: Any)

    func setIsHideCollidedSymbols(_ pathOverlay: NMFPath, rawFlag: Any)

    func getBounds(_ pathOverlay: NMFPath, success: (_ bounds: Dictionary<String, Any>) -> Void)
}

internal extension  PathOverlayHandler {
    func handlePathOverlay(pathOverlay: NMFOverlay,
                           method: String,
                           args: Any?,
                           result: @escaping FlutterResult) {
        let pathOverlay = pathOverlay as! NMFPath
        switch method {
        case NPathOverlay.coordsName: setCoords(pathOverlay, rawCoords: args!)
        case NPathOverlay.widthName: setWidth(pathOverlay, rawWidthDp: args!)
        case NPathOverlay.colorName: setColor(pathOverlay, rawColor: args!)
        case NPathOverlay.outlineWidthName: setOutlineWidth(pathOverlay, rawWidthDp: args!)
        case NPathOverlay.outlineColorName: setOutlineColor(pathOverlay, rawColor: args!)
        case NPathOverlay.passedColorName: setPassedColor(pathOverlay, rawColor: args!)
        case NPathOverlay.passedOutlineColorName: setPassedOutlineColor(pathOverlay, rawColor: args!)
        case NPathOverlay.progressName: setProgress(pathOverlay, rawProgress: args!)
        case NPathOverlay.patternImageName: setPatternImage(pathOverlay, rawNOverlayImage: args!)
        case NPathOverlay.patternIntervalName: setPatternInterval(pathOverlay, rawInterval: args!)
        case NPathOverlay.isHideCollidedCaptionsName: setIsHideCollidedCaptions(pathOverlay, rawFlag: args!)
        case NPathOverlay.isHideCollidedMarkersName: setIsHideCollidedMarkers(pathOverlay, rawFlag: args!)
        case NPathOverlay.isHideCollidedSymbolsName: setIsHideCollidedSymbols(pathOverlay, rawFlag: args!)
        case getterName(NPathOverlay.boundsName): getBounds(pathOverlay, success: result)
        default: result(FlutterMethodNotImplemented)
        }
    }
}
